File: Maths_And_Stats/Number_Theory/Sieve_of_Eratosthenes/Sieve_of_Erastosthenes.rb
class SieveOfErastosthenes
  def initialize(limit)
    @limit = limit
    @primes = initial_array
  end

  # Auto invokation of the class
  def self.call(limit)
    new(limit).call
  end

  def call
    find_primes
    @primes.select { |n| n }.length # Select only prime numbers in array and return only prime values.
  end
  
  private
  
  def initial_array
    (0..@limit).map {|i| i >= 2 ? true : false}
  end
  
  # Iterates over all truthies values to filter multiples and get primes
  def find_primes
    index = 0
    while index < @primes.length do
      find_multiples(index) if @primes[index]
      index += 1
    end
  end

  # Finds all multiples of the current index number, starting with its square.
  def find_multiples(index)
    multiples_index = index * index
    while multiples_index < @primes.length do
      @primes[multiples_index] = false
      multiples_index += index
    end
  end  
end

puts SieveOfErastosthenes.call(-1) # 0
puts SieveOfErastosthenes.call(1) # 0
puts SieveOfErastosthenes.call(100) # 25 
puts SieveOfErastosthenes.call(1000000) # 78498

File: Maths_And_Stats/Number_Theory/Sieve_of_Eratosthenes/Sieve_of_Erastosthenes.ts
const getAllPrimesUpToNumber = (num: number): number[] => {
  //Prefill array up to max num.
  const maxNumArray: boolean[] = new Array(num).fill(true)

  //Won't find any primes > sqrt of num.
  const limit = Math.floor(Math.sqrt(num))

  //Increment by 2, as no even numbers besides 2 are prime.
  for (let i = 3; i < limit; i += 2) {
    if (maxNumArray[i]) {
      //Remove multiples of primes.
      for (let j = i * i; j < num; j += i * 2) {
        maxNumArray[j] = false
      }
    }
  }

  //Initialize with 2, as it is the only even prime and we have not checked them.
  const primes: number[] = [2]

  //Push all items that are truthy to primes array.
  for (let i = 3; i < num; i += 2) {
    maxNumArray[i] && primes.push(i)
  }

  return primes
}

console.log(getAllPrimesUpToNumber(10000000))
